### ISSUE TITLE
Repo level concatenation of data

### ISSUE THREAD

**Issue by Bytes-Explorer** [deepseek-ai/DeepSeek-Coder#43]
Can you share more details on the technique for repo level concatenation part?

**Comment by guoday** [comment]
We first parsing the dependencies of files, e.g. A->B, B->C, B->D. And then we rearrange the file positions based on their dependencies, e.g. A,B,C,D. For file paths, we add them on each code file as a comment. An example is shown in https://github.com/deepseek-ai/DeepSeek-Coder#4-repository-level-code-completion

**Comment by Bytes-Explorer** [comment]
Thank you for your response. Is this done for all languages in the data?

**Comment by Bytes-Explorer** [comment]
Thank you @guoday

**Comment by Bytes-Explorer** [comment]
@guoday Do you then do [repo level dedup](https://github.com/deepseek-ai/DeepSeek-Coder/issues/42#issuecomment-1825826360) for all programming languages or just the above languages?

**Comment by guoday** [comment]
just the above languages. Other languages employ file level dedup.

**Comment by Bytes-Explorer** [comment]
@guoday Thank you for your prompt responses. I was curious if you did any ablation studies/evaluations to understand if repo level concatenation helped the model performance in a significant way.

**Comment by Bytes-Explorer** [comment]
Do you use HF benchmark for evaluations?

**Comment by guoday** [comment]
We will use public datasets like [RepoCoder](https://arxiv.org/pdf/2303.12570.pdf) and [CrossCodeEval](https://crosscodeeval.github.io/) to evaluate.

**Comment by Bytes-Explorer** [comment]
Ok thanks, was aware of those. Once again, appreciate your prompt responses. I look forward to reading the technical report from your group. Thanks!

**Comment by Casi11as** [comment]
![temp](https://github.com/deepseek-ai/DeepSeek-Coder/assets/63392846/5e57313e-e190-441f-be4f-7bd4c46f0171)

Hello, I would like to know the details of the concatenation of data. Assume that the structure of parsed dependencies is in the picture, what is the concatenation results? Is it ACF,ADF,ADG,BCF,BDF,BDG,BE?  7 pieces?

**Comment by guoday** [comment]
First, we select the file with the smallest incoming degree, and if there are multiple files with the smallest incoming degree, we randomly choose one. This process is repeated until a dependency order is obtained. For your example, there are many possibilities, one of which could be BACDFGE.

**Comment by Casi11as** [comment]
> First, we select the file with the smallest incoming degree, and if there are multiple files with the smallest incoming degree, we randomly choose one. This process is repeated until a dependency order is obtained. For your example, there are many possibilities, one of which could be BACDFGE.

In other words, will all the files of the same language in a repo only concatenate one sample?

**Comment by guoday** [comment]
Theoretically, yes. However, to shorten the sample length, we will parse a repository in advance and then divide it into multiple independent subgraphs based on dependencies, with each independent subgraph regarding as a sample.

**Comment by guoday** [comment]
The term "independent subgraph" refers to a weakly connected subgraph. First, convert the directed graph into an undirected graph, and then divide the graph into multiple connected subgraphs. That is, in each subgraph, any two vertices should be connected by edges within the subgraph. In your example, it is a connected subgraph, with only one subgraph, which is itself. The following is the code to divide the graph into subgraphs.

```python
# convert the directed graph into an undirected graph
def to_undirected(graph):
    undirected_graph = defaultdict(set)
    for node in graph:
        undirected_graph[node]
        for neighbor in graph[node]:
            undirected_graph[node].add(neighbor)
            undirected_graph[neighbor].add(node)
    return undirected_graph

# Use DFS to find all connected subgraphs.
def dfs(graph, node, visited, subgraph):
    visited[node] = True
    subgraph.add(node)
    for neighbor in graph[node]:
        if not visited[neighbor]:
            dfs(graph, neighbor, visited, subgraph)

# obtain all subgraphs
def get_subgraphs(graph):
    undirected_graph = to_undirected(graph)
    visited = {node: False for node in undirected_graph}
    subgraphs = []
    for node in undirected_graph:
        if not visited[node]:
            subgraph = set()
            dfs(undirected_graph, node, visited, subgraph)
            subgraphs.append(subgraph)
    return subgraphs
```

**Comment by guoday** [comment]
> Regarding repo-level concatenation, I have a related question.
> 
> In a batch, one sample may contain multi docs from different files, such as repo_a/file_a and repo_a/file_b. When concatenating these files into one sample for pre-training and calculating loss, will there still be an attention mask to prevent file_b from attending file_a?
> 
> If there is an attention mask, how can it serve the purpose of capturing the repository context? If there isn't, training by simply concatenating the beginning and end of different files seems somewhat peculiar.

If `file_b` depends on `file_a`, why is there a need for an attention mask to prevent `file_b` from focusing on `file_a`? Conversely, if `file_b` doesn't depend on `file_a`, we wouldn't concatenate these files into a single sample.

**Comment by guoday** [comment]
> > Regarding repo-level concatenation, I have a related question.
> > In a batch, one sample may contain multi docs from different files, such as repo_a/file_a and repo_a/file_b. When concatenating these files into one sample for pre-training and calculating loss, will there still be an attention mask to prevent file_b from attending file_a?
> > If there is an attention mask, how can it serve the purpose of capturing the repository context? If there isn't, training by simply concatenating the beginning and end of different files seems somewhat peculiar.
> 
> If `file_b` depends on `file_a`, why is there a need for an attention mask to prevent `file_b` from attending `file_a`? Conversely, if `file_b` doesn't depend on `file_a`, we wouldn't concatenate these files into a single sample.

In specific scenarios, such as the one described in https://github.com/deepseek-ai/DeepSeek-Coder/issues/43#issuecomment-1831433765, the input sequence is BACDFGE. In this case, even though file E does not directly depend on file A,  the file E  is allowed to attend the file A. This approach enables the model to effectively utilize contextual information from various files within the repository, thereby enhancing its overall comprehension and performance.

**Comment by slamandar** [comment]
Thank you for your prompt and detailed response! 

One last question. 

In one sample, is there a need for special token between the concatenated files? So that the model can distinguish that there are multiple files, and avoid model generate code like "import package" after the main content, in some downstream scenarios.

**Comment by guoday** [comment]
In fact, special token is required. However, we incorporate comments such as #utils.py and #model.py before each file to indicate to the model that the code completion is at the repository level.

**Comment by slamandar** [comment]
Completely understand. Thanks again for your quick response!

**Comment by Bytes-Explorer** [comment]
@guoday I was also wondering what do you do to the other files, like build files or metadata files? Thanks

**Comment by vaisaxena** [comment]
@guoday Thanks for the details above. It was quite helpful. One follow up question. 

Do you take care of cycles that may appear in the dependency graph of the files ? How do you handle that ?  This is the case where A->B, B->C, C->A

**Comment by dongs0104** [comment]
> > > Regarding repo-level concatenation, I have a related question.
> > > In a batch, one sample may contain multi docs from different files, such as repo_a/file_a and repo_a/file_b. When concatenating these files into one sample for pre-training and calculating loss, will there still be an attention mask to prevent file_b from attending file_a?
> > > If there is an attention mask, how can it serve the purpose of capturing the repository context? If there isn't, training by simply concatenating the beginning and end of different files seems somewhat peculiar.
> > 
> > 
> > If `file_b` depends on `file_a`, why is there a need for an attention mask to prevent `file_b` from attending `file_a`? Conversely, if `file_b` doesn't depend on `file_a`, we wouldn't concatenate these files into a single sample.
> 
> In specific scenarios, such as the one described in [#43 (comment)](https://github.com/deepseek-ai/DeepSeek-Coder/issues/43#issuecomment-1831433765), the input sequence is BACDFGE. In this case, even though file E does not directly depend on file A, the file E is allowed to attend the file A. This approach enables the model to effectively utilize contextual information from various files within the repository, thereby enhancing its overall comprehension and performance.

@guoday 
But Node `A` and `B` is not connected in-direct graph so i have question about it what if A and B contain simair contents? did you get `get_subgraphs` and then re order it by repo level again?

**Comment by guoday** [comment]
> > > > Regarding repo-level concatenation, I have a related question.
> > > > In a batch, one sample may contain multi docs from different files, such as repo_a/file_a and repo_a/file_b. When concatenating these files into one sample for pre-training and calculating loss, will there still be an attention mask to prevent file_b from attending file_a?
> > > > If there is an attention mask, how can it serve the purpose of capturing the repository context? If there isn't, training by simply concatenating the beginning and end of different files seems somewhat peculiar.
> > > 
> > > 
> > > If `file_b` depends on `file_a`, why is there a need for an attention mask to prevent `file_b` from attending `file_a`? Conversely, if `file_b` doesn't depend on `file_a`, we wouldn't concatenate these files into a single sample.
> > 
> > 
> > In specific scenarios, such as the one described in [#43 (comment)](https://github.com/deepseek-ai/DeepSeek-Coder/issues/43#issuecomment-1831433765), the input sequence is BACDFGE. In this case, even though file E does not directly depend on file A, the file E is allowed to attend the file A. This approach enables the model to effectively utilize contextual information from various files within the repository, thereby enhancing its overall comprehension and performance.
> 
> @guoday But Node `A` and `B` is not connected in-direct graph so i have question about it what if A and B contain simair contents? did you get `get_subgraphs` and then re order it by repo level again?

Nodes `A` and `B` are connected in an undirected graph, indicating that they are the same input sequence. If A and B have similar contents, B can leverage the content of A as additional context to enhance the completion process (sssuming in the sequence, B follows A). We do not re-order these nodes.

**Comment by reignianor** [comment]
Remarkable work! I wonder the reason of concatenation's benefits, do you first pretrain with file-level code, and then perform continual-train with repo-level code, and have y

**Comment by reignianor** [comment]
Truly remarkable work! I am curious about the advantages of repo concatenation in your training process. Do you first pre-train using file-level code, and then continue-train with repo-level code? What if pre-train using repo-level code from the start?

**Comment by zte-tcb** [comment]
> @guoday Thanks for the details above. It was quite helpful. One follow up question.
> 
> Do you take care of cycles that may appear in the dependency graph of the files ? How do you handle that ? This is the case where A->B, B->C, C->A

I have the same doubts.

**Comment by juncaofish** [comment]
> > @guoday Thanks for the details above. It was quite helpful. One follow up question.
> > Do you take care of cycles that may appear in the dependency graph of the files ? How do you handle that ? This is the case where A->B, B->C, C->A
> 
> I have the same doubts.

Actually, the dependencies of file within the same repository could be represented as a DAG? and it's impossible for the case as your show for A->B, B->C, C->A, which would cause a circular reference problem

**Comment by guoday** [comment]
> > > @guoday Thanks for the details above. It was quite helpful. One follow up question.
> > > Do you take care of cycles that may appear in the dependency graph of the files ? How do you handle that ? This is the case where A->B, B->C, C->A
> > 
> > 
> > I have the same doubts.
> 
> Actually, the dependencies of file within the same repository could be represented as a DAG? and it's impossible for the case as your show for A->B, B->C, C->A, which would cause a circular reference problem

The algorithm employs a modified topological sort. Unlike the standard approach that selects nodes with zero in-degrees, this algorithm selects nodes with minimal in-degrees, which allows it to handle cycles within the graph.

**Comment by slamandar** [comment]
In the process of reproducing repository-level data concatenation, I have a question. 

Is the file-level data or the unparsed language data(excluding python/java/c/c++/c#) included in the long-context continue pre-train dataset?

**Comment by kail8** [comment]
Hi, I'm really impressed by your advanced work. I have am extra question: in the repo-level concatenation, if one file depends on some huge files or libraries (such as `torch` or `transformers`), the concatenated sample will inevitably exceed the window size / context length. How do you deal with this problem?

**Comment by guoday** [comment]
> In the process of reproducing repository-level data concatenation, I have a question.
> 
> Is the file-level data or the unparsed language data(excluding python/java/c/c++/c#) included in the long-context continue pre-train dataset?

Yes. For unparsed language data or repo-level code that exceed 32k, we keep them in file-level format in the long-context continue pre-training.

**Comment by guoday** [comment]
> Hi, I'm really impressed by your advanced work. I have am extra question: in the repo-level concatenation, if one file depends on some huge files or libraries (such as `torch` or `transformers`), the concatenated sample will inevitably exceed the window size / context length. How do you deal with this problem? @guoday

For unparsed language data or repository-level code that surpasses 32KB, we split them into file-level data for use in the continue pre-training.

**Comment by kail8** [comment]
> > Hi, I'm really impressed by your advanced work. I have am extra question: in the repo-level concatenation, if one file depends on some huge files or libraries (such as `torch` or `transformers`), the concatenated sample will inevitably exceed the window size / context length. How do you deal with this problem? @guoday
> 
> For unparsed language data or repository-level code that surpasses 32KB, we split them into file-level data for use in the continue pre-training.

Got it. Thanks for your quick response!

**Comment by Calvinnncy97** [comment]
May I know how are the dependencies parsed?

**Comment by virtualzx-nad** [comment]
hi @guoday !  when i use the model how do i structure my repo in my prompt to take advantage DeepSeek's understanding of repo structures?   How should i separate different files in the same repo and how do i denote filenames?   My repo also contains different languages so just adding `# filename.py` doesn't seem to be good enough.

**Comment by tangbo-sh** [comment]
The technical report shows that only 8B data is used to train the long context, while the readme shows that 200B is trained on the 16K window. So I am a little confused here. If the pretrain truncation does have 200B data for 16K window training, then when was the base frequency and scaling factor of the rope modified? Thank you.

Content in the technical report: "The model underwent an additional
1000 steps of training, using a batch size of 512 and a sequence length of 16K. "
